napari/layers/utils/_color_encoding.py

```diff
@@ -0,0 +1,257 @@
+        The safe constant fallback color to use if the feature column
+        does not contain valid color values.
+    """
+
+    encoding_type: Literal['DirectColorEncoding'] = 'DirectColorEncoding'
+    feature: str
+    fallback: ColorValue = Field(default_factory=lambda: DEFAULT_COLOR)
+
+    def __call__(self, features: Any) -> ColorArray:
+        # A column-like may be a series or have an object dtype (e.g. color names),
+        # neither of which transform_color handles, so convert to a list.
+        return ColorArray.validate_type(list(features[self.feature]))
+
+
+class NominalColorEncoding(_DerivedStyleEncoding[ColorValue, ColorArray]):
+    """Encodes color values from a nominal feature whose values are mapped to colors.
+
+    Attributes
+    ----------
+    feature : str
+        The name of the feature that contains the nominal values to be mapped to colors.
+    colormap : CategoricalColormap
+        Maps the feature values to colors.
+    fallback : ColorValue
+        The safe constant fallback color to use if mapping the feature values to
+        colors fails.
+    """
+
+    encoding_type: Literal['NominalColorEncoding'] = 'NominalColorEncoding'
+    feature: str
+    colormap: CategoricalColormap
+    fallback: ColorValue = Field(default_factory=lambda: DEFAULT_COLOR)
+
+    def __call__(self, features: Any) -> ColorArray:
+        # map is not expecting some column-likes (e.g. pandas.Series), so ensure
+        # this is a numpy array first.
+        values = np.asarray(features[self.feature])
+        return self.colormap.map(values)
+
+
+class QuantitativeColorEncoding(_DerivedStyleEncoding[ColorValue, ColorArray]):
+    """Encodes color values from a quantitative feature whose values are mapped to colors.
+
+    Attributes
+    ----------
+    feature : str
+        The name of the feature that contains the nominal values to be mapped to colors.
+    colormap : Colormap
+        Maps feature values to colors.
+    contrast_limits : Optional[Tuple[float, float]]
+        The (min, max) feature values that should respectively map to the first and last
+        colors in the colormap. If None, then this will attempt to calculate these values
+        from the feature values the first time this generate color values. If that attempt
+        fails, these are effectively (0, 1).
+    fallback : ColorValue
+        The safe constant fallback color to use if mapping the feature values to
+        colors fails.
+    """
+
+    encoding_type: Literal[
+        'QuantitativeColorEncoding'
+    ] = 'QuantitativeColorEncoding'
+    feature: str
+    colormap: Colormap
+    contrast_limits: Optional[Tuple[float, float]] = None
+    fallback: ColorValue = Field(default_factory=lambda: DEFAULT_COLOR)
+
+    def __call__(self, features: Any) -> ColorArray:
+        values = features[self.feature]
+        if self.contrast_limits is None:
+            self.contrast_limits = _calculate_contrast_limits(values)
```

Comment:
I wonder whether, rather than having "call" set state on the encoding, we should instead *always* calculate contrast limits on "call" if they are not set? ie out of all functions in this work, "call" should be imho the most stateless? And indeed, rather than having the encoding take in a table and refer to state, could it be stateless and always take in a series or array?

Reply:
Good catch and agreed. I brought this work in from an older PR, which had a slightly different design and this was an oversight on my part. I'll address this before we merge here.

Reply:
Done. We will need to handle refreshing the contrasting limits if we want to implement napari's current behavior, but I think that's better done in a PR that actually uses this (e.g. using `ColorEncoding` in `Points`).